ash/src/prelude.rs

```diff
@@ -19,3 +19,27 @@ impl vk::Result {
         }
     }
 }
+
+/// Repeatedly calls `f` until it does not return [`vk::Result::INCOMPLETE`] anymore,
+/// ensuring all available data has been read into the vector.
+///
+/// See for example [`vkEnumerateInstanceExtensionProperties`]: the number of available
+/// items may change between calls; if it has increased `INCOMPLETE` is returned because
+/// the result-vec is not large enough.
+///
+/// [`vkEnumerateInstanceExtensionProperties]: https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/vkEnumerateInstanceExtensionProperties.html
+pub(crate) unsafe fn read_into_vector<T>(
+    f: impl Fn(&mut u32, *mut T) -> vk::Result,
+) -> VkResult<Vec<T>> {
+    loop {
+        let mut count: u32 = 0;
+        f(&mut count, std::ptr::null_mut()).result()?;
+        let mut data = Vec::with_capacity(count as usize);
```

Comment:
Do we want to keep this vector around in subsequent loop iterations and use `.reserve()` instead, likely using just a `realloc` instead of `free` + `malloc`? OTOH we don't need the original memory to be copied, that's probably making it less efficient in the end.

Reply:
Nothing will be copied since the length is zero, so this is fine to do, though I don't think it'll matter in practice.

Reply:
The `Vec<>` len will be zero, but the length of the allocation will not - `realloc` looks at the previous size passed to `malloc` which is based on what was passed to `with_capacity`, it doesn't know about `Vec<>` having a counter determining how much is actually initialized with useful data (leaving detailed optimizations of ie. pages being physically allocated only when they're actually going to be written to aside).

This is just a moot point anyways, let's resolve and not care about it.